Support Newmark-Beta time integration in FSI ALE simulation

lindsayad opened the pull request:
Refs #31248

lindsayad committed:
Initial user input

lindsayad committed:
Add space in DynamicSolidMechanicsPhysics md header to match QuasiStatic

lindsayad committed:
Add static method for computing Newmark-Beta vel/accel in InertialForce

I'll leverage this in the FSI velocity continuity class

lindsayad committed:
Create ADPenaltyVelocityContinuityNewmarkBeta

Refs #31248

lindsayad committed:
Leverage new class in user's ALE input

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3126728/), step Docs: sync website on c781b72 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31278/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3126778/), step Generate coverage on c781b72 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Fsi

<table>
  <thead>
    <tr>
      <th></th>
      <th>80a80d</th>
      <th colspan=4>#31278 c781b7</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.56%</td>
      <td>80.15%</td>
      <td>-8.42%</td>
      <td>11.36%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>333</td>
      <td>331</td>
      <td>-2</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>43</td>
      <td>82</td>
      <td>+39</td>
      <td>39</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31278/coverage_diff/fsi)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31278/coverage/fsi)

### Solid mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>80a80d</th>
      <th colspan=4>#31278 c781b7</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.01%</td>
      <td>85.01%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>28068</td>
      <td>28070</td>
      <td>+2</td>
      <td>7</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4951</td>
      <td>4951</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31278/coverage_diff/solid_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31278/coverage/solid_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31278/coverage/xfem)

</details>

## Warnings

- `fsi` new line coverage rate 11.36% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3126778/), step Verify coverage on c781b72 wanted to post the following:


The following coverage requirement(s) failed:
- `fsi` coverage rate 80.15% is less than the required 85.0%

lindsayad committed:
Create test

lindsayad committed:
Make component parameter required

moosebuild commented:
Job [Non unity build](https://civet.inl.gov/job/3134168/) on 01694a4 : invalidated by @lindsayad

lindsayad committed:
Hopefully make nonAD FSI test more robust in the test harness

moosebuild commented:
Job [CUDA GPU](https://civet.inl.gov/job/3135277/) on 448c886 : invalidated by @lindsayad

maxnezdyur reviewed (changes requested):
The feature looks good in general. A few minor mistakes in the input file. The HHT alpha method is very powerful, I want to make sure it will not be prevented from being used.

lindsayad commented on modules/fsi/test/tests/2d-small-strain-transient/fsi_flat_channel.i:
```diff
@@ -294,7 +294,7 @@
 [Preconditioning]
   [./SMP]
-    type = SMP
+    type = FDP
```
There are occasionally sporadic test failures related to this input. One occurred in https://github.com/idaholab/moose/pull/30987#issuecomment-3190133986

lindsayad commented on modules/fsi/src/interfacekernels/ADPenaltyVelocityContinuityNewmarkBeta.C:
```diff
@@ -0,0 +1,77 @@
+  params.addRequiredParam<Real>("gamma", "gamma parameter for Newmark-Beta time integration");
+  return params;
+}
+
+ADPenaltyVelocityContinuityNewmarkBeta::ADPenaltyVelocityContinuityNewmarkBeta(
+    const InputParameters & parameters)
+  : ADPenaltyVelocityContinuity(parameters),
+    _beta(getParam<Real>("beta")),
+    _gamma(getParam<Real>("gamma"))
```
I dunno, do you think I should change the class name if we do?

lindsayad replied:
I think we can definitely add support for this in the future

lindsayad commented on modules/fsi/src/interfacekernels/ADPenaltyVelocityContinuity.C:
```diff
@@ -47,25 +47,27 @@ ADPenaltyVelocityContinuity::ADPenaltyVelocityContinuity(const InputParameters &
+  ADRealVectorValue ret;
+  for (const auto i : index_range(_solid_velocities))
+  {
+    mooseAssert(_solid_velocities[i], "This should be non-null");
```
You're right there's no need for this

lindsayad committed:
Address Max review

- Directly use dynamic kernels instead of using deprecated action
- Remove erroneous assert
- Fix mistake in InertialForce velocity param

maxnezdyur reviewed (approved):
This looks good to me now!
